python: enable all linting checks and type-hint the code

rmuir opened the pull request:
Review https://github.com/apache/lucene/pull/14322 separate as it is included here to prevent merge conflicts.

So look at the 2nd commit if you want to see what has changed:
* enable all type checking
* type-hint the code to get checker flowing
* disable a few remaining checks with a TODO list
* enable all linter checks
* autofix all safe fixes (e.g. api migrations)
* disable just the remaining checks with a massive TODO list

stefanvodita commented on dev-tools/scripts/create_line_file_docs.py:
```diff
@@ -60,7 +60,7 @@ def compress_with_seek_points(file_name_in, file_name_out, num_seek_points):
 
       bytes_in_chunk += len(line)
-      f_out.write(line)  # false positive in python's crazy typing # pyright: ignore[reportArgumentType]
+      f_out.write(line)
```
Curious - what changed that that is no longer necessary?

rmuir replied:
IIRC, it seems like a corner-case bug in the typechecker. typechecker has a grid of options: https://docs.basedpyright.com/latest/configuration/config-files/#diagnostic-settings-defaults

And in the process of getting started here, lots of booleans were being flipped on and off. I think it led to a false positive given the exact circumstances of booleans (ignore this problem, but fail on that problem, etc): basically we are asking it to validate within a broken/incoherent type system. 

Once enough stuff got fixed, then strictness was enabled which allows things to work smoothly. This is just my read on what happened.

rmuir replied:
In this example I'm pretty sure that because `file_name_out` parameter was untyped (Unknown), it made `gzip.open()` call more ambiguous, causing `f_out` to have type of `GzipFile | TextIOWrapper[_WrappedBuffer]`

Once `file_name_out` got a type of `str`, it reduces ambiguity and narrows result of `gzip.open()` and hence `f_out` to just be `GzipFile`.

You can see this yourself if you go in editor and change `file_name_out` to be Any

stefanvodita replied:
Got it, thank you for the explanation!

stefanvodita commented on dev-tools/scripts/pyproject.toml:
```diff
@@ -1,21 +1,163 @@
 
 # disabling/enabling of rules
 ignore = [
+  # This is a big TODO list of current linter problems.
```
Should we open an issue to make it more likely someone finds this and decides to carry on the work?

rmuir replied:
Yes I was planning to make a "good first issue" in both repositories for this? I thought first, we might want to have a look at the list and see if there is anything we are against, and simply move those further down in the file, away from the TODO list, instead under "list of rules we keep disabled". 

I am sure there are probably things here that at least aren't priorities: policing print() statements, TODOs, and docstrings are examples. Maybe we don't even want to adopt python naming conventions. These are just examples.

So we could remove any bikesheds from  the list first: it would be great to get some of these issues fixed such as variable shadowing, exception handling, etc

stefanvodita replied:
That's a good idea!

stefanvodita commented on dev-tools/scripts/releaseWizard.py:
```diff
@@ -887,7 +873,7 @@ def get_release_version():
   except Exception:
     print("Not a valid version %s" % v)
-    return get_release_version()
+    raise
```
This change doesn't look like the others.

rmuir replied:
I think some of the change was automatic. What else should we do if supplied with an invalid version? try to proceed anyway? :)

rmuir replied:
I know when battling this one, it was difficult to fix the issues because of situations like this. It causes helper function `get_release_version()` to have unnecessarily wide type (e.g. allowing None or maybe even other types in some places). And callsites are not checking for those exceptional cases, which makes for tons of chaos.

Better for exceptional cases to be exceptions?

stefanvodita replied:
It's hard for me to judge this one without knowing how it's used, but I can imagine it's frustrating to stop a long workflow with this exception because of a typo instead of prompting the user a second time. Now, from looking at the code, it seems that is not the case, so it should be fine to proceed.

rmuir committed:
python: enable all linting checks and type-hint the code

enable all type checking
type-hint the code to get checker flowing
disable a few remaining checks with a TODO list
enable all linter checks
autofix all safe fixes (e.g. api migrations)
disable just the remaining checks with a massive TODO list

stefanvodita reviewed (commented):
Reviewed the 2nd commit. Only small comments.

rmuir commented:
By the way, thank you for reviewing these changes! I know it isnt fun, and I know i'll cause something to break: but these scripts are important (releases, backwards-compatibility, community stuff) and I'll help fix any short-term fallout.

The changes make the scripts easier to work with too besides just linting. You get things like autocomplete and hover documentation in your editor when the types work, its 2025 :)

stefanvodita reviewed (approved):
Happy I could help review and fingers crossed it doesn't break anything too badly!
